core/trino-main/src/main/java/io/trino/server/security/oauth2/OAuth2Authenticator.java

```diff
@@ -49,22 +48,23 @@ public OAuth2Authenticator(OAuth2Service service, OAuth2Config config)
     @Override
     protected Optional<Principal> extractPrincipalFromToken(String token)
     {
-        Map<String, Object> claims = service.convertTokenToClaims(token);
-        if (claims.containsKey(principalField)) {
-            Optional<Map<String, Object>> claims = service.convertTokenToClaims(token);
-            return Optional.ofNullable(claims.get(principalField))
+        try {
+            return service.convertTokenToClaims(token)
+                    .map(claims -> claims.get(principalField))
                     .map(String.class::cast)
                     .map(BasicPrincipal::new);
         }
-        return Optional.empty();
+        catch (ChallengeFailedException e) {
+            return Optional.empty();
+        }
     }
 
     @Override
     protected AuthenticationException needAuthentication(ContainerRequestContext request, String message)
     {
         UUID authId = UUID.randomUUID();
-        URI redirectUri = service.startRestChallenge(request.getUriInfo().getBaseUri().resolve(CALLBACK_ENDPOINT), authId);
```

Comment:
Is changing the protocol needed to solve the "correctness issues"?

Reply:
This does not change the protocol.  It simply changes the initial URL, but the protocol is completely unchanged.

Reply:
Correct, and we should have always done that.  It lets us use a nonce to protect the authentication code.  Also in the future, we can use this to timeout clients faster... basically if a client doesn't hit the initiate in a few seconds, we kill the auth because likely the browser did not launch.

@11xor6 in the commit message mention the new use of nonce cookies for query clients.